packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -191,13 +201,29 @@ export class MonacoWorkspace implements lang.Workspace {
         this.textModelService.onDidCreate(model => this.fireDidOpen(model));
     }
 
-    protected _rootUri: string | null = null;
+    protected updateWorkspaceFolders(newRootDirs: FileStat[]): void {
+        const oldWorkspaceUris = this.workspaceFolders.map(folder => folder.uri.toString());
+        const newWorkspaceUris = newRootDirs.map(folder => folder.uri);
+        const added = newWorkspaceUris.filter(uri => oldWorkspaceUris.indexOf(uri) < 0).map((dir, index) => this.toWorkspaceFolder(dir, index));
+        const removed = oldWorkspaceUris.filter(uri => newWorkspaceUris.indexOf(uri) < 0).map((dir, index) => this.toWorkspaceFolder(dir, index));
+        this._workspaceFolders = newWorkspaceUris.map(this.toWorkspaceFolder);
+        this.onDidChangeWorkspaceFoldersEmitter.fire({ added, removed });
+    }
+
+    protected toWorkspaceFolder(uriString: string, index: number): WorkspaceFolder {
+        return {
+            uri: Uri.parse(uriString),
+            name: uriString.substring(uriString.lastIndexOf('/')) || uriString,
+            index
+        };
+    }
+
     get rootUri(): string | null {
-        return this._rootUri;
+        return this._workspaceFolders[0] && this._workspaceFolders[0].uri.toString();
```

Comment:
I think this incorrect, same for `rootPath`. This will never ever return with `null` but `undefined` silently. Am  I overlooking something?

```ts
const array = [];
function toUpper(): string | null {
    return array[0] && array[0].toLocaleLowerCase();
}
console.log(undefined === toUpper()); // true
console.log(null === toUpper()); // false
```
(You can try it with `ts-node` quickly.)

Reply:
Replaced